### PR TITLE
Submissions EDM from Howard

### PR DESCRIPTION
**Pull request by bluemellophone**

### CODE REVIEW

File: app/modules/organizations/models.py [hunk 1]
```diff
@@ -5,25 +5,124 @@
...
     Organizations database model.
     """
 
     guid = db.Column(
         db.GUID, default=uuid.uuid4, primary_key=True
     )  # pylint: disable=invalid-name
+    version = db.Column(db.Integer, default=None, nullable=True)
     title = db.Column(db.String(length=50), nullable=False)
+    # todo why can dateTimes in the auth class be non Nullable but in here the upgrade fails?
+    createdDate = db.Column(
```
Comment (bluemellophone) [inline app/modules/organizations/models.py:107]: A `HoustonModel` inherits from `FeatherModel` which has a parent `TimestampViewed` which inherits from `Timestamp`.  The models `TimestampViewed` and `Timestamp` create the database columns `viewed` and `created`, respectively.  These columns can be deleted since it is implicit.

File: app/modules/organizations/models.py [hunk 2]
```diff
@@ -5,25 +5,124 @@
...
+        db.DateTime, default=datetime.datetime.now(tz=TIMEZONE), nullable=True
+    )
+    modifiedDate = db.Column(
+        db.DateTime, default=datetime.datetime.now(tz=TIMEZONE), nullable=True
+    )
+
+    logoGuid = db.Column(db.GUID, default=uuid.uuid4, nullable=True)
```
Comment (bluemellophone) [inline app/modules/organizations/models.py:114]: Please use `_` delimiters for attributes, not camel case in order to adhere to PEP8.

File: migrations/versions/56ce9dfdbf4f_.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import sqlalchemy_utils
+
+import app
+import app.extensions
+
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    with op.batch_alter_table('organization', schema=None) as batch_op:
```
Comment (bluemellophone) [inline migrations/versions/56ce9dfdbf4f_.py:24]: This update will need to be redone after the created, viewed, and camel case attribute names are fixed

**Review by bluemellophone — COMMENTED** [review]
Small changes requested to database

**Review by bluemellophone — COMMENTED** [review]
LGTM

**Review by bluemellophone — COMMENTED** [review]
LGTM

**Comment by bluemellophone** [comment]
@hwindsor This PR is ready for review.  It runs all tests successfully and merges with the latest code on our develop branch.

File: app/modules/organizations/models.py [hunk 4]
```diff
@@ -79,20 +80,33 @@ def _process_members(self, members):
             #    self.members.append(user)
```
Comment (hwindsor) [inline app/modules/organizations/models.py:80]: This bit also needs to be restored now you have the relationship working, this is the part that stitches them together on the sync

File: app/modules/organizations/models.py [hunk 5]
```diff
@@ -103,26 +117,21 @@ class Organization(db.Model, HoustonModel, OrganizationEDMMixin):
...
+    members = db.relationship('OrganizationUserMemberships', back_populates='membership')
+
     def __repr__(self):
         return (
             '<{class_name}('
             'guid={self.guid}, '
             'title="{self.title}", '
             'website={self.website}, '
-            'logoUrl={self.logoUrl}, '
-            #'members={self.members}'
```
Comment (hwindsor) [inline app/modules/organizations/models.py]: I was going to add this back in once the sync worked. It was helpful. Why did you remove it?

File: migrations/versions/dcecb0cd0cea_.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
+"""empty message
```
Comment (hwindsor) [inline migrations/versions/dcecb0cd0cea_.py:1]: Any reason you added two migrations rather than one?

File: tests/modules/submission/resources/test_ensure_submission.py [hunk 7]
```diff
@@ -33,7 +33,7 @@ def test_ensure_clone_submission_by_uuid(
             '4d46c55d-accf-29f1-abe7-a24839ec1b95'
         )
 
-        assert clone.temp_submission.commit_houston_api_version == '0.1.0.8b208226'
+        # assert clone.temp_submission.commit_houston_api_version == '0.1.0.8b208226'
```
Comment (hwindsor) [inline tests/modules/submission/resources/test_ensure_submission.py:36]: Yes. I wanted to talk about this one, this was specifically not something I added, it was already in the test. how do you think we should manage what the required houston_api_version should be in future

File: tasks/app/__init__.py [hunk 8]
```diff
@@ -5,40 +5,40 @@
 
 from invoke import Collection
 
-from . import (
```
Comment (hwindsor) [inline tasks/app/__init__.py:8]: Is there some logic about the order that the imports and the Collection is created? 
If there is, a comment would help to prevent future errors.

**Comment by hwindsor** [comment]
Hi JP, 
You've made a few substantial changes.
Created the new class for bidirectional relationship between User and Org,
Changed the module order in config.py and in the tasks/app/__init__.py
Moved a lot of the imports of module.*.model classes from file scope to function scope.

I'm happy that these have fixed the issue but it's not clear to me what the problem was with the code as it was and which of these changes solves the problem.
I had tried changing the order of users and orgs in config.py and it didn't fix the problem for me.
Should we avoid file scope module.*.models imports in future? What's the problem that this is fixing?

File: tasks/app/__init__.py (hunk 8 above)
Comment (bluemellophone) [reply tasks/app/__init__.py:8]: No, the order here does not matter.  What matters is that the modules we import do not import objects in app/models/ out of order.  The way to ensure this is to move all module-level imports to function-level imports.  We want the order of the modules to be defined in `config.py` because we have database ORM dependencies.  For example if we import `app.modules.assets.models.Asset` for a test, it may be that it is imported before the `submission` module, breaking the import structure we have setup.  The order of these imports shouldn't matter.

File: migrations/versions/dcecb0cd0cea_.py (hunk 6 above)
Comment (bluemellophone) [reply migrations/versions/dcecb0cd0cea_.py:1]: No special reason.  We will merge our database updates down the line once we have a more stable version of the backend.

Comment (bluemellophone) [reply migrations/versions/dcecb0cd0cea_.py]: I have merged the database updates into one migration

File: app/modules/organizations/models.py (hunk 5 above)
Comment (bluemellophone) [reply app/modules/organizations/models.py]: I have pushed code to re-enable `Organization.members`

File: app/modules/organizations/models.py (hunk 4 above)
Comment (bluemellophone) [reply app/modules/organizations/models.py]: This is a bit more complicated now.   We store the relationship as a `HoustonModel` with its own separate database table so that things like created and updated timestamps are preserved.  There is a new test in `tests/modules/organizations/test_models.py` that shows how to add memberships now.  I'd also request you take a look at how the properties for `User.memberships` and `Organization.members` is defined.

**Comment by bluemellophone** [comment]
I have pushed a few fixes:

- Re-enable `Organization.members` and `User.memberships`, please review
- Consolidated the database migrations
- Added a new test to show how organization memberships are created

File: tests/modules/submission/resources/test_ensure_submission.py (hunk 7 above)
Comment (bluemellophone) [reply tests/modules/submission/resources/test_ensure_submission.py]: I found the bug for this test and fixed it in commit dddea84

File: app/modules/organizations/models.py [hunk 9]
```diff
@@ -5,28 +5,145 @@
...
+        return organization, is_new
+
+    def _process_members(self, members):
+        from app.modules.users.models import User
+
+        for member in members:
+            log.info('Adding Member ID %s' % (member.id,))
+            user, is_new = User.ensure_edm_obj(member.id)
+
+            # todo restore once the tests work
```
Comment (hwindsor) [inline app/modules/organizations/models.py:78]: The pedant in me says this line should be removed but the realist says this needs to be pulled without any further delay.

Comment (bluemellophone) [reply app/modules/organizations/models.py]: Removed!

**Review by naknomum — APPROVED** [review]
yeah i think this looks sound to me.  i am sure i will have to catch up on in a bit more to incorporate into my enc_passthru feature branch, but that will be a good way for me to _make sure_ i am caught up.  also, i am guessing the db migration stuff might have diverged a little, but we can check that out and fix when we run it.
